### ISSUE TITLE
Tailscale breaks WSL2 DNS

### ISSUE THREAD

**Issue by PieterScheffers** [tailscale/tailscale#2815]
On Windows login, Tailscale is started. When I open Ubuntu 18.04 on WSL2 no DNS is working.
```
$ dig google.com
; <<>> DiG 9.11.3-1ubuntu1.12-Ubuntu <<>> google.com
;; global options: +cmd
;; connection timed out; no servers could be reached
```

When opening `/etc/wsl.conf`, I see Tailscale has added the following snippet:
```
# added by tailscale
[network]
generateResolvConf = false
```

When I remove this snippet from `/etc/wsl.conf` and then terminate and start WSL again, the DNS works again.
```
C:\WINDOWS\system32> wsl --terminate Ubuntu-18.04
C:\WINDOWS\system32> wsl --distribution Ubuntu-18.04

$ dig google.com
; <<>> DiG 9.11.3-1ubuntu1.12-Ubuntu <<>> google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 54109
;; flags: qr rd ad; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;google.com.                    IN      A

;; ANSWER SECTION:
google.com.             0       IN      A       142.250.179.206

;; Query time: 21 msec
;; SERVER: 172.29.64.1#53(172.29.64.1)
;; WHEN: Wed Sep 08 09:16:49 CEST 2021
;; MSG SIZE  rcvd: 54
```

On every restart of Windows the snippet is added to `/etc/wsl.conf` again.

**Comment by DentonGentry** [comment]
The write to /etc/wsl.conf was added in https://github.com/tailscale/tailscale/commit/1147c7fd4f1cdb2eb9727cb34a5af0bf09ca84f7 to fix https://github.com/tailscale/tailscale/issues/775.

When first booted with "generateResolvConf = false" present, what is the content of /etc/resolv.conf? I think we'd expect to have written a "nameserver 100.100.100.100" line there.

**Comment by PieterScheffers** [comment]
The `/etc/resolv.conf` file doesn't exist when the DNS isn't working.

**Comment by DentonGentry** [comment]
In https://login.tailscale.com/admin/dns do you have either Magic DNS enabled, or any nameservers added?

**Comment by PieterScheffers** [comment]
I see MagicDNS is enabled and nameserver 100.100.100.100 is the nameserver. The global nameserver is 1.1.1.1.

**Comment by DentonGentry** [comment]
So we should have written to both /etc/wsl.conf and /etc/resolv.conf, but something went wrong in writing to /etc/resolv.conf and the result is broken DNS.

Could you send the Tailscale IP address of the Windows host to support@tailscale.com, and reference this bug? Maybe we can see what went wrong in telemetry or error it reported.

**Comment by lorengordon** [comment]
I'm seeing the same thing in my setup.

**Comment by rubendibattista** [comment]
Same problem here. MagicDNS enabled, 1.1.1.1 as global nameserver

**Comment by bradms** [comment]
Likewise. I wondered what was causing this, and it turned out to be WSL2. MagicDNS enabled, and 1.1.1.1 as global nameserver (same as person above me)

**Comment by lorengordon** [comment]
The upgrade to 1.14.4 re-broke it, by re-adding `generateResolvConf = false` to `/etc/wsl.conf`.

**Comment by bradms** [comment]
> I have found Tailscale has also completely broken WSL DNS for me (under Ubuntu), but I didn't see any changes to my `/etc/resolv.conf` made by Tailscale.
> 
> It's worth noting that I am only running Tailscale on the Windows host, not in the WSL virtual machine.
> 
> The only reason I know it is the cause of the breakage is because I completely uninstalled Tailscale and restarted, and everything started working again.

I wonder if the latest update solved this issue? Either way, I am being cautious about it.

**Comment by gruberdev** [comment]
> The upgrade to 1.14.4 re-broke it, by re-adding `generateResolvConf = false` to `/etc/wsl.conf`.

Seems plausible since

> As a temporary workaround, I ran the command `sudo chattr +i /etc/resolv.conf` to prevent it from being modified. This is less than ideal. I do understand what they are trying to do, but it does not work, unfortunately. :/

Fixed it temporarily to me. I'm on a Windows 10 AME 1.61.1 with WSL2 and using Debian. 

> > I have found Tailscale has also completely broken WSL DNS for me (under Ubuntu), but I didn't see any changes to my `/etc/resolv.conf` made by Tailscale.
> > It's worth noting that I am only running Tailscale on the Windows host, not in the WSL virtual machine.
> > The only reason I know it is the cause of the breakage is because I completely uninstalled Tailscale and restarted, and everything started working again.
> 
> I wonder if the latest update solved this issue? Either way, I am being cautious about it.

When you mention new updates, do you refer to 1.61.x? Because if you do, I do have bad news. This issue hasn't been addressed so far.

**Comment by wbkang** [comment]
`sudo chattr +i /etc/resolv.conf` is pretty good workaround for now.

This is pretty annoying bug - it happens every restart. While Tailscale the company figures out solution, could you at least provide us an escape hatch to disable this behaviour? I don't see this behaviour documented anywhere other than here so that was surprising.

**Comment by lorengordon** [comment]
Still broken with 1.18.0.

The `chattr` option didn't work for me. I ended up adding a line to my profile to create `/etc/resolv.conf`:

```
[ -f /etc/resolv.conf ] || sudo echo "nameserver $(ip route show default | awk '{print $3}')" > /etc/resolv.conf
```

**Comment by neiltwist** [comment]
I too have this issue on 1.18.0

> The write to /etc/wsl.conf was added in [1147c7f](https://github.com/tailscale/tailscale/commit/1147c7fd4f1cdb2eb9727cb34a5af0bf09ca84f7) to fix #775.
> 
> When first booted with "generateResolvConf = false" present, what is the content of /etc/resolv.conf? I think we'd expect to have written a "nameserver 100.100.100.100" line there.

There is no resolv.conf file, but a symbolic link to `../run/resolvconf/resolv.conf`. What is the content of the file supposed to be?

**Comment by stealthybox** [comment]
When I'm on my home WiFi, using nslookup on `100.100.100.100` works for looking up tailnet devices AND internet hostnames like `google.com`.

For the cafe WiFi I'm using right now, the DNS lookups on `100.100.100.100` for tailnet machines still work, but `google.com` does not.
Instead, it just hangs until it times out.

I had to add `timeout:1` and a fallback `nameserver 1.1.1.1` to my resolv.conf to work around it.

Perhaps this is because the new WiFi network in my Windows network preferences doesn't have an overridden DNS server, and the tailscale service is using the wrong API(s) to deduce what the local DNS settings should be? (just guessing)

On https://login.tailscale.com/admin/dns, I have "Override Local DNS" disabled, but I do have 1.0.0.1 and 1.1.1.1 listed after Local DNS settings.

It's worth noting that the WSL2 default gateway is the DNS server that gets auto-generated in a default WSL2 resolv.conf
You can also use that as a fallback server for the Window's hosts Local DNS settings.
Calculate the address like so:
```bash
ip route | grep default | awk '{print $3}'
```

**Comment by arjoh** [comment]
I'm having the same issue with 1.18.2

**Comment by arjoh** [comment]
> I'm having the same issue with 1.18.2

Just a minute after posting this I started to work on the workaround proposed by @lorengordon above, and after disconnecting Tailscale and reconnecting again things actually started to work as expected.

When connecting Tailscale, resolv.conf is generated by Tailscale:

```
# resolv.conf(5) file generated by tailscale
# DO NOT EDIT THIS FILE BY HAND -- CHANGES WILL BE OVERWRITTEN

nameserver 100.100.100.100
search ...
```

And after disconnecting:
```
# This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
# [network]
# generateResolvConf = false
nameserver...
```

wsl.conf remains the same:
```
# added by tailscale
[network]
generateResolvConf = false
```

**Comment by lorengordon** [comment]
I still have 1.18.0 and also noticed the behavior seems to have changed in my last reboot. I expected to be prompted by sudo so it could write the resolv.conf, but I wasn't. Checked it, and saw the same content as @arjoh. But it's hard to tell if it is fixed, or a fluke of some kind, as 1.18.0 *wasn't* working at first. And I haven't seen a linked commit or PR to this issue, and no commentary from the tailscale team.

**Comment by lorengordon** [comment]
Hmm, I wonder if the problem is because in wsl2 `/etc/resolv.conf` is typically a symlink to `../run/resolvconf/resolv.conf`? It being a symlink is why I had to modify my original workaround from:

```
[ -e /etc/resolv.conf ] || sudo bash -c "echo 'nameserver $(ip route show default | awk "{print $3}")' > /etc/resolv.conf"
```
to:
```
[ -h /etc/resolv.conf -a ! -e /etc/resolv.conf ] && sudo bash -c "rm /etc/resolv.conf; echo 'nameserver $(ip route show default | awk "{print $3}")' > /etc/resolv.conf"
```

But once I did that, `/etc/resolv.conf` was no longer a symlink. I bet *then* the tailscale logic was able to work.

**Comment by lorengordon** [comment]
Hysterically, now the backup of the file that tailscale makes is invalid, because wsl2 changes its IP every boot. 🤦‍♂️

**Comment by voltagex** [comment]
Please allow disabling this - you're breaking WSL2 even when Tailscale isn't connected. 

`ping: tailscale.com: Temporary failure in name resolution`

**Comment by voltagex** [comment]
Also, don't change files that aren't yours without asking me first.

**Comment by evilpilaf** [comment]
I was running into the same issue and after updating to tailscale 1.18.2 I was able to remove `/etc/wsl.conf` and now my DNS is working as expected.

**Comment by Xalaxis** [comment]
I'm unsure what version of Tailscale exactly did it, but this bug has broken networking on all of my WSL2 machines at some point. Deleting the `/etc/wsl.conf` file seems to fix the issue across Tailscale restarts, so maybe the fix/removal has been added to newer versions?

**Comment by DentonGentry** [comment]
In the 1.20.x release https://github.com/tailscale/tailscale/commit/a14d445fc7220c7c827a93a8a8c4970b45cf449c means that resolv.conf will only be written if a TS_DEBUG_CONFIGURE_WSL environment variable exists.

**Comment by arjoh** [comment]
> In the 1.20.x release [a14d445](https://github.com/tailscale/tailscale/commit/a14d445fc7220c7c827a93a8a8c4970b45cf449c) means that resolv.conf will only be written if a TS_DEBUG_CONFIGURE_WSL environment variable exists.

This is currently not working for me. Set TS_DEBUG_CONFIGURE_WSL to True in Windows 11 System environment variables. Tailscale version is 1.20.4.

**Comment by DentonGentry** [comment]
Fixed in 1.20 by https://github.com/tailscale/tailscale/pull/3687

We may make another attempt to better support WSL in the future. I do not believe this bug will be useful in that, we won't do so until we have something better. Closing this bug.

**Comment by Rican7** [comment]
### EDIT: Warning the comment below pertains to WSL1. This ticket is titled WSL2. The two stacks are very different.


I just ran into this myself, and I wanted to report what I ran into (and how **I fixed it**) in-case it helps future work on WSL compatibility or anyone else that finds this thread.

## Problem

I had just recently installed Tailscale, and restarted my machine, and then I was running into connectivity (really DNS) issues under WSL (WSL 1, Ubuntu 22.04).

I checked my `/etc/resolv.conf` file, as that's a common source of DNS issues in WSL (microsoft/WSL#3438), and saw this:

```
$ cat /etc/resolv.conf
# This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
# [network]
# generateResolvConf = false
nameserver fec0:0:0:ffff::1
nameserver fec0:0:0:ffff::2
nameserver fec0:0:0:ffff::3
search tail63fee.ts.net Home
```

<details>
<summary>... which resulted in DNS failing, but I could still ping</summary>
<pre><code>
$ dig google.com
;; communications error to fec0:0:0:ffff::1#53: timed out
;; communications error to fec0:0:0:ffff::1#53: timed out
;; communications error to fec0:0:0:ffff::1#53: timed out
;; communications error to fec0:0:0:ffff::2#53: timed out
;; communications error to fec0:0:0:ffff::3#53: timed out

; <<>> DiG 9.18.18-0ubuntu0.22.04.2-Ubuntu <<>> google.com
;; global options: +cmd
;; no servers could be reached

$ ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=118 time=3.79 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=118 time=3.50 ms
64 bytes from 8.8.8.8: icmp_seq=3 ttl=118 time=3.41 ms
64 bytes from 8.8.8.8: icmp_seq=4 ttl=118 time=3.13 ms
64 bytes from 8.8.8.8: icmp_seq=5 ttl=118 time=3.28 ms
64 bytes from 8.8.8.8: icmp_seq=6 ttl=118 time=3.52 ms
64 bytes from 8.8.8.8: icmp_seq=7 ttl=118 time=3.56 ms
^C
--- 8.8.8.8 ping statistics ---
7 packets transmitted, 7 received, 0% packet loss, time 6005ms
rtt min/avg/max/mdev = 3.133/3.455/3.791/0.195 ms
</code></pre>
</details>

... Exiting the Tailscale app (in the system tray), removing the resolv.conf (`sudo rm /etc/resolv.conf` in WSL), and then restarting WSL (`wsl --shutdown` via an admin/elevated `cmd`) resulted in this `/etc/resolv.conf` contents:

```
$ cat /etc/resolv.conf
# This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
# [network]
# generateResolvConf = false
nameserver fec0:0:0:ffff::1
nameserver fec0:0:0:ffff::2
nameserver fec0:0:0:ffff::3
search Home
```

... which is _close_, but notice the `search` line is missing the tailscale domain?


## Fix

1. I removed the resolv.conf again (`sudo rm /etc/resolv.conf` in WSL)
2. I shutdown WSL (`wsl --shutdown` via an admin/elevated `cmd`)
3. I disabled the "Tailscale" (Tailscale Tunnel) network adapter in Network Connections
![explorer_2024-02-22_18-51-27](https://github.com/tailscale/tailscale/assets/742384/3e97a8d6-a9b9-40ba-8c25-46431b6bf377)
4. Started WSL again...

<details><summary>... and now my DNS in WSL is fixed!</summary>
<pre><code>
$ cat /etc/resolv.conf
# This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
# [network]
# generateResolvConf = false
nameserver 192.168.0.1
nameserver 205.171.3.25
nameserver 2001:428::1
search Home
$ dig google.com

; <<>> DiG 9.18.18-0ubuntu0.22.04.2-Ubuntu <<>> google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 12874
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
; COOKIE: 8dc1cce36c2a16b30100000065d7fa3fef29bcd1cacc18e0 (good)
;; QUESTION SECTION:
;google.com.                    IN      A

;; ANSWER SECTION:
google.com.             188     IN      A       142.250.72.14

;; Query time: 4 msec
;; SERVER: 192.168.0.1#53(192.168.0.1) (UDP)
;; WHEN: Thu Feb 22 18:51:59 MST 2024
;; MSG SIZE  rcvd: 83

$ ping google.com
PING google.com(den08s06-in-x0e.1e100.net (2607:f8b0:400f:803::200e)) 56 data bytes
64 bytes from den08s06-in-x0e.1e100.net (2607:f8b0:400f:803::200e): icmp_seq=1 ttl=116 time=4.23 ms
64 bytes from den08s06-in-x0e.1e100.net (2607:f8b0:400f:803::200e): icmp_seq=2 ttl=116 time=3.54 ms
64 bytes from den08s06-in-x0e.1e100.net (2607:f8b0:400f:803::200e): icmp_seq=3 ttl=116 time=3.78 ms
64 bytes from den08s06-in-x0e.1e100.net (2607:f8b0:400f:803::200e): icmp_seq=4 ttl=116 time=3.86 ms
64 bytes from den08s06-in-x0e.1e100.net (2607:f8b0:400f:803::200e): icmp_seq=5 ttl=116 time=4.00 ms
^C64 bytes from 2607:f8b0:400f:803::200e: icmp_seq=6 ttl=116 time=3.71 ms

--- google.com ping statistics ---
6 packets transmitted, 6 received, 0% packet loss, time 25043ms
rtt min/avg/max/mdev = 3.538/3.853/4.233/0.220 ms
</code></pre>
</details>

I hope someone else finds this useful!

**Comment by derekyle** [comment]
Adding to this for those that have the same issue as me. The above solution didn't work for me because every time I tried to disable the Tailscale tunnel, it would go away and a new connection would instantly start up and be enabled. Instead, I went to properties and unchecked the IPv6 option and it seems to have fixed the DNS problem inside WSL for me. No deletions or resolv.conf or anything else required. Did not even need a restart of WSL.
